modeshape-jcr/src/main/java/org/modeshape/jcr/locking/StandaloneLockingService.java

```diff
@@ -0,0 +1,222 @@
+ *  are hard-referenced. Therefore for locks which are in use, the key will be hard-referenced by the value of the entry as well.
+ *  - once a lock in unlocked, the strong key reference will be nulled-out so that the GC is free to reclaim the entry from a map
+ *  - when multiple threads are contending for the same lock, they will reestablish the strong reference so that it's not GCed 
+ * </p>
+ * 
+ * @author Horia Chiorean (hchiorea@redhat.com)
+ * @since 5.0
+ */
+@ThreadSafe
+public class StandaloneLockingService implements LockingService {
+
+    private static final Logger LOGGER = Logger.getLogger(StandaloneLockingService.class);
+
+    private final Map<String, NamedLock> locksByName = new WeakHashMap<>();
+    private volatile long lockTimeoutMillis = 0;
+
+    public StandaloneLockingService() {
+    }
+
+    @Override
+    public boolean tryLock(String... names) {
+        return tryLock(lockTimeoutMillis, TimeUnit.MILLISECONDS, names);
+    }
+
+    @Override
+    public void setLockTimeout(long lockTimeoutMillis) {
+        CheckArg.isNonNegative(lockTimeoutMillis, "lockTimeoutMillis");
+        this.lockTimeoutMillis = lockTimeoutMillis;
+    }
+
+    @Override
+    public boolean tryLock(long time, TimeUnit unit, String... names) {
+        Set<String> successfullyLocked = new HashSet<>();
+        for (String name : names) {
+            LOGGER.debug("Attempting to lock {0}", name);
+            // search for an existing lock or create a new one atomically...
+            NamedLock lock = locksByName.get(name);
+            if (lock == null) {
+                synchronized (locksByName) {
+                    lock = locksByName.computeIfAbsent(name, NamedLock::new);
+                }
+            }
+            boolean success = false;
+            try {
+                success = time > 0 ? lock.tryLock(time, unit) : lock.tryLock();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Thread " + Thread.currentThread().getName()
+                             + " received interrupt request while waiting to acquire lock '{0}'", name);
+                Thread.currentThread().interrupt();
+            }
+            if (!success) {
+                LOGGER.debug("Unable to acquire lock on {0}. Reverting back the already obtained locks: {1}", name,
+                             successfullyLocked);
+                synchronized (locksByName) {
+                    successfullyLocked.stream().forEach(lockName -> locksByName.get(lockName).unlock());
+                }
+                return false;
+            }
+
+            // we now have a lock but we need to make sure it hasn't just been marked for cleanup
+            // this may've been unlocked by another thread, so refresh to make sure a strong reference is active
+            validateLock(name, lock);
+
+            LOGGER.debug("{0} locked successfully", lock);
+            successfullyLocked.add(name);
+        }
```

Comment:
As mentioned in [line comments](https://github.com/hchiorean/modeshape/commit/c96d4ebfd2c8a18df13f3a433a8a2e2d4b092703#commitcomment-15155073) on one of the commits, I think this method should:

1) hold onto the acquired locks via a set (see how the `ClusteringService` does it in this same PR)
1) if any lock cannot be acquired, any already-acquired locks can be unlocked directly via this set, saving the need to look up the locks by name again on [line 103](https://github.com/ModeShape/modeshape/pull/1496/files#diff-5ba93ccae783c9c45dec04404e62e328R103)

In the other commit's comments, I suggested moving the `validateLock` methods into the `NamedLock`, but that won't necessarily be cleaner. So leaving those methods where they are is fine with me.